### ISSUE TITLE
Unify code for parallel uploades in S3/azure/gcp storages

### ISSUE THREAD

**Issue by marpio** [gomods/athens#255]
At the moment each Saver uploades .zip,.mod and .info in parallel but does this a little bit different.
It would be good to use common code for that.

**Comment by robjloranger** [comment]
I agree, they are all pretty similar but it would be nice to have one implementation to keep things clean. Some thing maybe at the (./pkg/storage) level.

**Comment by marpio** [comment]
I would like to work on this.

**Comment by robjloranger** [comment]
@marpio If you don't mind including https://github.com/gomods/athens/pull/208#discussion_r202346292 we can probably get the gcp stuff in and make merge conflicts less complicated.

**Comment by marpio** [comment]
@robjloranger sure!
